psydac/linalg/tests/test_linalg.py

```diff
@@ -823,6 +833,63 @@ def test_operator_evaluation():#n1, n2, p1, p2, P1=False, P2=False):
     H = H1 @ ( H2 + H3 - H4 + H5 ).T
     assert np.linalg.norm((H@v).toarray() - v.toarray(), ord=2) < 10 * tol
 
+#===============================================================================
+
+def test_internal_storage():
+
+    # Create LinearOperator Z = A @ A.T @ A {@ A.T}, where the domain and codomain of A are of different dimension.
+    # Prior to a fix, operator would not have enough preallocated storage defined.
+
+    n1=2
+    n2=1
+    p1=1
+    p2=1
+    P1=False
+    P2=False
+    V = get_StencilVectorSpace(n1, n2, p1, p2, P1, P2)
+    U1 = BlockVectorSpace(V, V)
+    U2 = BlockVectorSpace(V, V, V)
+
+    x1 = StencilVector(V)
+    x1[0] = 1
+    x1[1] = 1
+    x = BlockVector(U1, (x1, x1))
+    xx = BlockVector(U2, (x1, x1, x1))
+
+    A1 = StencilMatrix(V, V)
+    A1[0, 0, 0, 0] = 1
+    A1[1, 0, 0, 0] = 1
+    A = BlockLinearOperator(U1, U2, ((A1, A1), (A1, A1), (A1, A1)))
+    B = A.T
+    C = A
+    D = A.T
+
+    Z1_1 = A @ (B @ C)
+    Z1_2 = (A @ B) @ C
+    Z1_3 = A @ B @ C
+    y1_1 = Z1_1 @ x
+    y1_2 = Z1_2 @ x
+    y1_3 = Z1_3 @ x
+
+    Z2_1 = (A @ B) @ (C @ D)
+    Z2_2 = (A @ B @ C) @ D
+    Z2_3 = A @ (B @ C @ D)
+    Z2_4 = A @ B @ C @ D
+    y2_1 = Z2_1 @ xx
+    y2_2 = Z2_2 @ xx
+    y2_3 = Z2_3 @ xx
+    y2_4 = Z2_4 @ xx
+
+    assert len(Z1_1.tmp_vectors) == 2
+    assert len(Z1_2.tmp_vectors) == 2
+    assert len(Z1_3.tmp_vectors) == 2
+    assert len(Z2_1.tmp_vectors) == 3
+    assert len(Z2_2.tmp_vectors) == 3
+    assert len(Z2_3.tmp_vectors) == 3
+    assert len(Z2_4.tmp_vectors) == 3
+    assert np.all( y1_1.toarray() == y1_2.toarray() ) & np.all( y1_2.toarray() == y1_3.toarray() )
+    assert np.all( y2_1.toarray() == y2_2.toarray() ) & np.all( y2_2.toarray() == y2_3.toarray() ) & np.all( y2_3.toarray() == y2_4.toarray() )
```

Comment:
The standard way to compare Numpy arrays for exact equality is using the function [`array_equal`](https://numpy.org/doc/stable/reference/generated/numpy.array_equal.html). Here you should also avoid calling `toarray()` on the same object:
```suggestion
a1_1, a1_2, a1_3       = [v.toarray() for v in (y1_1, y1_2, y1_3)]
a2_1, a2_2, a2_3, a2_4 = [v.toarray() for v in (y2_1, y2_2, y2_3, y2_4)]
assert np.array_equal(a1_1, a1_2) and np.array_equal(a1_2, a1_3)
assert np.array_equal(a2_1, a2_2) and np.array_equal(a2_2, a2_3) and np.array_equal(a2_3, a2_4)
```

Reply:
Thanks for the feedback!